docker/tarfile/src.go

```diff
@@ -43,28 +42,31 @@ type layerInfo struct {
 //       the source of an image.
 // 	To do for both the NewSourceFromFile and NewSourceFromStream functions
 
-// NewSourceFromFile returns a tarfile.Source for the specified path
-// NewSourceFromFile supports both conpressed and uncompressed input
+// NewSourceFromFile returns a tarfile.Source for the specified path.
 func NewSourceFromFile(path string) (*Source, error) {
 	file, err := os.Open(path)
 	if err != nil {
 		return nil, errors.Wrapf(err, "error opening file %q", path)
 	}
 	defer file.Close()
 
-	reader, err := gzip.NewReader(file)
+	// If the file is already not compressed we can just return the file itself
+	// as a source. Otherwise we pass the stream to NewSourceFromStream.
+	stream, isCompressed, err := compression.AutoDecompress(file)
 	if err != nil {
+		return nil, errors.Wrapf(err, "detecting compression for file %q", path)
+	}
+	if !isCompressed {
 		return &Source{
 			tarPath: path,
 		}, nil
 	}
-	defer reader.Close()
```

Comment:
So, it turns out that `gzip.NewReader` returns a `ReadCloser`, and the caller is expected to really `Close` the stream; the `compression.DetectCompression` implementation missed that.

Luckily(?) the `gzip.Reader.Close()` is (currently?) trivial, it only returns an error value which `Read()` would AFAICT would have returned anyway as long as the consumer is reading until EOF.

But that complicates the `AutoDecompress` design; should it wrap uncompressed inputs in `NopCloser`? Should it return a separate `close` callback instead of modifying the `io.Reader` stream (like `dirImageMockWithRef` does)?

Or do we just pretend that `gzip.Reader.Close` does not exist? :)

Reply:
We could change the API to return an `io.ReadCloser`. Since `io.Reader` ⊆ `io.ReadCloser` there shouldn't be a problem with any users of the interface (aside from the downside they probably won't notice they should call `Close` now).

Reply:
… and that
```go
reader = …
defer reader.Close()
reader, … = AutoDecompress(…)
defer reader.Close() // Again!
```
is unintuitive.  But then none of the alternatives I can think of are any more elegant.

Reply:
Maybe, but if you compare it to `gzip.NewReader` you have to do a similar thing.

```
reader := ...
defer reader.Close()
reader2, ... := gzip.NewReader(reader)
defer reader2.Close()
```

We could work around it by doing some very dodgy `.(io.Closer)` handling in order to still allow users to pass `io.Reader`s that don't have a hidden `Close()` method. But that's probably what you were thinking when you said that you can't think of any more elegant workarounds. :wink:

Reply:
The unintuitive part, to me, is that with raw `gzip`, there is `reader.Close()` and `reader`**`2`**`.Close()`; with `AutoDecompress`, there would be two `reader.Close()`s, which looks pretty similar to a copy&paste bug.

But I still can’t think of a much better API.
```go
reader = …
defer reader.Close()
reader, close, … = AutoDecompress(…)
defer close()
```
does not look noticeably better.

Reply:
For the moment I'm just going to ignore the close method for `gzip` if that's okay with you. Otherwise we'd have to `NopCloser` most of the other functions, and `DetectCompression`'s signature would look wrong.

Reply:
I’m not really happy to ignore resource cleanup and error detection, no. It’s a time bomb, and doing the right thing seems plausible.

I can take on the work of updating the existing `compression` package and its users, if you don’t want to deal with the rest of c/image as part of this PR (apart from updating it to use the new API).

Reply:
Alright, in that case I'll work on doing that tomorrow.

Reply:
I'll work on fixing this up today.